File: crates/parelthon_models/src/comment.rs
//
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::point2d::Point2D;

#[derive(Debug, sqlx::FromRow, Deserialize, Serialize)]
pub struct VideoComment {
    pub comment_id: Uuid,
    pub user_id: Option<Uuid>,
    pub video_id: Option<Uuid>,
    pub start_time: f32,
    pub end_time: Option<f32>,
    // pub coordinates: Point2D,
    pub screen_x: f32,
    pub screen_y: f32,
    pub comment_text: String,
    pub updated_at: chrono::DateTime<chrono::Utc>,
    pub created_at: chrono::DateTime<chrono::Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateVideoComment {
    pub video_id: Option<Uuid>,
    pub user_id: Option<Uuid>,
    pub comment_text: String,
    pub coordinates: Point2D,
    pub start_time: f32,
    pub end_time: Option<f32>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct GetVideoComment {
    pub comment_id: Uuid,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct UpdateVideoComment {
    pub comment_id: Uuid,
    pub markdown: String,
}

File: clients/react-app/src/models/comment.ts
import { z } from 'zod';

const createVideoCommentSchema = z.object({
    video_id: z.union([z.string().uuid(), z.null()]).optional(),
    user_id: z.union([z.string().uuid(), z.null()]).optional(),
    comment_text: z.string(),
    coordinates: z.object({
        x: z.number(),
        y: z.number(),
    }),
    start_time: z.number(),
    end_time: z.union([z.number(), z.null()]).optional(),
});


const videoCommentSchema = z.object({
    comment_id: z.string().uuid(),
    user_id: z.union([z.string().uuid(), z.null()]).optional(),
    video_id: z.union([z.string().uuid(), z.null()]).optional(),
    start_time: z.number(),
    end_time: z.union([z.number(), z.null()]).optional(),
    updated_at: z.string().datetime(),
    screen_x: z.number(),
    screen_y: z.number(),
    comment_text: z.string(),
    created_at: z.string().datetime(),
});



type VideoComment = z.infer<typeof videoCommentSchema>;

type CreateVideoComment = z.infer<typeof createVideoCommentSchema>;


export type { VideoComment, CreateVideoComment };

